File: benchmarks/executable/float_integer_rounding_runtime.c
// Expected cases (user arguments => exit; stdout; stderr):
// [] => 0; "Rounded 42\n"; ""
// ["x"] => 1; ""; ""
// ["x", "y"] => 1; ""; ""

#include <inttypes.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

_Static_assert(sizeof(double) == sizeof(uint64_t),
               "binary64 requires 64-bit double storage");

static double from_bits(uint64_t bits) {
  double value;
  memcpy(&value, &bits, sizeof(value));
  return value;
}

int main(int argc, char **argv) {
  (void)argv;
  const uint64_t bits = argc <= 1 ? UINT64_C(0x4045600000000000)
                                  : UINT64_C(0x7ff8000000000000);
  const double value = from_bits(bits);
  if (!isfinite(value) || value < -128.0 || value >= 128.0) return 1;
  const int8_t rounded = (int8_t)value;
#ifdef _WIN32
  if (_setmode(_fileno(stdout), _O_BINARY) == -1) return 1;
#endif
  if (printf("Rounded %" PRId8 "\n", rounded) < 0) return 1;
  return 0;
}

File: benchmarks/executable/float_integer_rounding_runtime.rs
// Expected cases (user arguments => exit; stdout; stderr):
// [] => 0; "Rounded 42\n"; ""
// ["x"] => 1; ""; ""
// ["x", "y"] => 1; ""; ""

fn main() {
    let bits = if std::env::args_os().skip(1).next().is_none() {
        0x4045_6000_0000_0000_u64
    } else {
        0x7ff8_0000_0000_0000_u64
    };
    let value = f64::from_bits(bits);
    if !value.is_finite() || !(-128.0..128.0).contains(&value) {
        std::process::exit(1);
    }
    let rounded = value as i8;
    println!("Rounded {rounded}");
}
